### PR TITLE
let the owner restore files if they where moved out of a shared folder

### PR DESCRIPTION
**Pull request by schiessle**
add a copy to the owners trash bin if another user moves a file out of a shared folder

cc @nickvergessen

### CODE REVIEW

File: apps/files_trashbin/lib/Storage.php [hunk 1]
```diff
@@ -58,8 +67,38 @@ function __construct($parameters, IUserManager $userManager = null) {
...
+
+		if ($currentUser) {
+			$currentUserId = $currentUser->getUID();
+
+			$view = new View($currentUserId . '/files');
+			$sourceStorage = $view->getFileInfo($oldPath)->getStorage();
+			$sourceOwner = $view->getOwner($oldPath);
+			$targetOwner = $view->getOwner($newPath);
+
+			if($sourceOwner !== $targetOwner
```
Comment (icewind1991) [inline apps/files_trashbin/lib/Storage.php]: comparing `$view->getFileInfo($oldPath)->getStorage() !== $view->getFileInfo($newPath)->getStorage() && $sourceStorage->instanceOfStorage(SharedStorage::class)` will be more robust in cases where you move between 2 shares.

File: apps/files_trashbin/lib/Storage.php [hunk 2]
```diff
@@ -58,8 +67,38 @@ function __construct($parameters, IUserManager $userManager = null) {
...
+				&& $sourceStorage->instanceOfStorage(SharedStorage::class)
```
Comment (icewind1991) [inline apps/files_trashbin/lib/Storage.php]: iirc using `SharedStorage::class` will throw an error when the sharing app is disabled, using the string directly should be fine

**Review by icewind1991 — CHANGES_REQUESTED** [review]

**Comment by icewind1991** [comment]
What happens if both users try to restore their separate copies?

**Comment by schiessle** [comment]
> What happens if both users try to restore their separate copies?

The file is only in the trashbin the owner, so only he can restore it. If the owner restores it and later the share recipient tries to move it back he will get the typical conflict dialog. If the recipient first move it back and then the owner restores it the restored file will be names "file (restored).txt"

**Review by icewind1991 — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Pull request merged**